### ISSUE TITLE
Implementation of LCA algorithm does not match with referenced paper

### ISSUE THREAD

**Issue by dtekinoglu** [networkx/networkx#5547]
The [reference paper](https://www3.cs.stonybrook.edu/~bender/pub/JALG05-daglca.pdf) stated in NetworkX documentation for lowest common ancestor, describes the algorithm as follow:

> Preprocess G by partitioning the set of edges into two sets: S1 is the set of edges of a spanning tree T of G, and S2
> is composed of the remaining edges, which make up the DAG D = (V, S2). Label the vertices of T by depth, breaking
> ties arbitrarily. The same labels are used for D. Create a list Lv for each vertex v, which contains the ancestors of v in
> D, including v itself. Proceed from vertex 0 until vertex n − 1. Sort the lists according to the first appearance of their
> nodes in an Euler tour of T .

However, when I look at the current implementation, I saw that there are major differences.

- NetworkX implementation sorts the lists with respect to Euler tour of G although it should be of T.
```
euler_tour_pos = {}
for node in nx.depth_first_search.dfs_preorder_nodes(G, root):
    if node not in euler_tour_pos:
        euler_tour_pos[node] = next(counter)

```
- The paper states that Lv lists contain ancestors of v in D. However, in NetworkX implementation, Lv contains ancestors of v in G.

```
    ancestors = {}
    for v in dag:
        if pairs is None or v in pairset:
            my_ancestors = nx.ancestors(G, v)
            my_ancestors.add(v)
            ancestors[v] = sorted(my_ancestors, key=euler_tour_pos.get)
```

IMO, current implementation does not follow the referenced study. Besides, the code seems complicated and difficult to understand. As it is stated in [Issue 4458](https://github.com/networkx/networkx/issues/4458), it does not work properly in some corner cases. I could not succesfully track the root cause down since there is a confict between the reference and the current implementation. 

I would suggest to implement the algorithm using [Euler Tour + RMQ method](https://discuss.codechef.com/t/tutorial-lowest-common-ancestor-using-euler-tour-range-minimum-query/94326) which is more readable and not worse than the current one in terms of time and space complexity.

@rossbar @MridulS @dschult

**Comment by dschult** [comment]
Sorry this took so long to respond to... I didn't catch it earlier.

The LCA code is suspect. See #4942 for a fairly long discussion that lead to us changing the `ancestors` call you point out to a call on `G` instead of `dag`.  That fixed the bug but we never understood why. It made sense to me why it would work, but I couldn't make sense of the paper describing the algorithm. Others tried to find a workaround, another paper, something that would make this function match a known algorithm. 

 In the end @jamestrimble posted some code for a [naive version of the function](https://github.com/networkx/networkx/issues/4942#issuecomment-957337481) and said that he didn't have the time to put it into a PR.  He also has a cryptic comment that there is a link on a wikipedia page to a paper with essentially this algorithm. 

If you would like to wrap that code into a PR that adds that method to the `lowest_common_ancestor.py` module, add tests, etc, that would be great!  The next step would be to figure out if and why the results are different. I suspect we want to replace the current existing code with this simpler method that is apparently also much faster. We can help of course -- and we will get to it eventually if you don't do it.

**Comment by dtekinoglu** [comment]
Thanks @dschult , this looks like a great opportunity for me. Since I am also working on a nx-guides notebook on LCA, working on what you've suggested would help a lot.

**Comment by U-238** [comment]
@dtekinoglu please also see #5628 which contains sample data that triggers an incorrect result

**Comment by dschult** [comment]
This can be closed due to #5736 and #5883
Thanks!
